### PR TITLE
add example for tensorflow

### PR DESCRIPTION
**Pull request by shenkejie15**
增加了 tensorflow mnist 的例子。

有一个已知的问题，tensorflow 2.8.0 使用 CPU 训练时有较高概率在训练完成后报错 core dumped，因此这里使用了 GPU。使用的队列是 a100，使用的资源是 CPU: 16, Memory: 8Gi, tensorstack.dev/nvidia-gpu: 2。这么设置是为了在 nc201 顺利运行，但不利于迁移。

### CODE REVIEW

**Comment by shenkejie15** [comment]
使用这里的代码进行训练、继续训练、评估的例子参考 https://proxy.nc201.kube.tensorstack.net/t9k/asset-hub/web/#/models/3623abde-b06d-4c44-9e3c-024f3a675268/280551d6-bca5-447a-b753-6e5e5ca56655/?tab=training（用户为 demo）

**Comment by ppkube** [comment]
> 有一个已知的问题，tensorflow 2.8.0 使用 CPU 训练时有较高概率在训练完成后报错 core dumped，因此这里使用了 GPU

不要使用 GPU，可以换换 tf 的版本试试

File: mnist-tensorflow/workflow-train.yaml [hunk 1]
```diff
@@ -0,0 +1,100 @@
+apiVersion: batch.tensorstack.dev/v1beta1
+kind: WorkflowTemplate
+metadata:
+  name: train-tensorflow
+spec:
+  params:
+    - name: command
+    - name: working_dir
+    - name: image
+    - name: request_cpu
```
Comment (ppkube) [inline mnist-tensorflow/workflow-train.yaml]: 这些 resource 相关的 params 都去掉，直接放到 yaml 里

File: mnist-tensorflow/workflow-train.yaml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  labels:
+    batch.tensorstack.dev/workflowTemplate: train-tensorflow
+spec:
+  params:
+    - name: command
+      value: "python asset-hub-example/mnist-tensorflow/train.py --log_dir ./log --dataset_dir dataset --save_path saved_model/model_state_dict.pt"
+    - name: working_dir
+      value: /t9k/mnt
+    - name: image
+      value: registry.tensorstack.cn/t9kmirror/tensorflow:2.8.0
```
Comment (ppkube) [inline mnist-tensorflow/workflow-train.yaml]: use a newer version of tensorflow?

File: mnist-tensorflow/workflow-continue.yaml [hunk 3]
```diff
@@ -0,0 +1,100 @@
+apiVersion: batch.tensorstack.dev/v1beta1
+kind: WorkflowTemplate
+metadata:
+  name: train-tensorflow
+spec:
+  params:
+    - name: command
+    - name: working_dir
+    - name: image
+    - name: request_cpu
```
Comment (ppkube) [inline mnist-tensorflow/workflow-continue.yaml]: remove resource related params

File: mnist-tensorflow/workflow-continue.yaml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      fieldSelector: status.phase==Failed
+    manifest: |
+      apiVersion: batch.tensorstack.dev/v1beta1
+      kind: TensorFlowTrainingJob
+      metadata:
+        generateName: model-retrain-
+      spec:
+        runPolicy:
+          cleanUpPolicy: None
+          backoffLimit: 0           # 所有Pod最多共重启0次
```
Comment (ppkube) [inline mnist-tensorflow/workflow-continue.yaml]: remove comments

File: mnist-tensorflow/workflow-continue.yaml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        scheduler:
```
Comment (ppkube) [inline mnist-tensorflow/workflow-continue.yaml]: remove scheduler config

File: mnist-tensorflow/workflow-continue.yaml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+spec:
+  params:
+    - name: command
+    - name: working_dir
+    - name: image
+    - name: request_cpu
+    - name: request_memory
+    - name: limit_cpu
+    - name: limit_memory
+    - name: worker_num
```
Comment (ppkube) [inline mnist-tensorflow/workflow-continue.yaml]: remove worker_num

File: mnist-tensorflow/workflow-evaluate.yaml [hunk 7]
```diff
@@ -0,0 +1,92 @@
+apiVersion: batch.tensorstack.dev/v1beta1
+kind: WorkflowTemplate
+metadata:
+  name: evaluate-tensorflow
+spec:
+  params:
+    - name: command
+    - name: working_dir
+    - name: image
+    - name: request_cpu
```
Comment (ppkube) [inline mnist-tensorflow/workflow-evaluate.yaml]: ditto

File: mnist-tensorflow/workflow-evaluate.yaml [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+      fieldSelector: status.phase==Failed
+    manifest: |
+      apiVersion: batch.tensorstack.dev/v1beta1
+      kind: TensorFlowTrainingJob
+      metadata:
+        generateName: model-evaluate-
+      spec:
+        runPolicy:
+          cleanUpPolicy: Unfinished
+          backoffLimit: 5           # 所有Pod最多共重启20次
```
Comment (ppkube) [inline mnist-tensorflow/workflow-evaluate.yaml]: ditto

File: mnist-tensorflow/workflow-evaluate.yaml [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        replicaSpecs:
+          - type: worker
+            replicas: 1
+            restartPolicy: OnFailure
+            template:
+              spec:
+                securityContext:
+                  runAsUser: 1000
```
Comment (ppkube) [inline mnist-tensorflow/workflow-evaluate.yaml:36]: explain why 1000 is used. That tensorflow image sets user = 1000?

File: mnist-tensorflow/workflow-evaluate.yaml [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  labels:
+    batch.tensorstack.dev/workflowTemplate: evaluate-tensorflow
+spec:
+  params:
+    - name: command
+      value: "python asset-hub-example/mnist-tensorflow/evaluate.py --no_cuda --dataset_dir dataset --load_path model/model_state_dict.pt --output_dir results"
+    - name: working_dir
+      value: /t9k/mnt
+    - name: image
+      value: registry.tensorstack.cn/t9kmirror/tensorflow:2.8.0
```
Comment (ppkube) [inline mnist-tensorflow/workflow-evaluate.yaml]: ditto

File: mnist-tensorflow/workflow-train.yaml [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+            queue: a100
+            priority: 10
+        replicaSpecs:
+          - type: worker
+            replicas: $(params.worker_num)
+            restartPolicy: Never
+            template:
+              spec:
+                securityContext:
+                  runAsUser: 1000
```
Comment (ppkube) [inline mnist-tensorflow/workflow-train.yaml:36]: ditto

File: mnist-tensorflow/evaluate.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    logger.setLevel(logging.INFO)
+    logger.addHandler(logging.StreamHandler())
+    logger.propagate = False
+
+    model = models.load_model(args.load_path)
+
+    dataset_dir=os.path.join(args.dataset_dir,'MNIST','raw')
+    test_images, test_labels = mnist_to_numpy(data_dir=dataset_dir, train=False)
+
+    test_images = test_images.reshape((10000, 28, 28, 1))
```
Comment (ppkube) [inline mnist-tensorflow/evaluate.py:86]: explain why these numbers are used

**Review by ppkube — CHANGES_REQUESTED** [review]

File: mnist-tensorflow/workflow-evaluate.yaml (hunk 9 above)
Comment (shenkejie15) [reply mnist-tensorflow/workflow-evaluate.yaml:36]: 和 tensorflow 无关。

Asset Hub launcher workflow 使用 t9k/build-sdk 镜像执行下载 Git 代码、下载 Asset、上传 Asset 的任务。t9k/build-sdk 使用的 uid 是 1000。

File: mnist-tensorflow/workflow-train.yaml (hunk 2 above)
Comment (shenkejie15) [reply mnist-tensorflow/workflow-train.yaml]: 改成了最新的 2.11.0，但是现在集群中 trainingjob 可能无法使用，所以没有实际测试过。

File: mnist-tensorflow/workflow-continue.yaml (hunk 6 above)
Comment (shenkejie15) [reply mnist-tensorflow/workflow-continue.yaml]: fixed
